designs/odcr.md

```diff
@@ -0,0 +1,408 @@
+
+#### Only launch ODCR instances
+
+```yaml
+apiVersion: karpenter.sh/v1
+kind: NodePool
+metadata:
+ name: default
+spec:
+ requirements:
+ - key: karpenter.sh/capacity-type
+   operator: In
+   values: ["reserved"]
+```
+
+#### Launch ODCR instances with on-demand fallback
+
+```yaml
+apiVersion: karpenter.sh/v1
+kind: NodePool
+metadata:
+ name: default
+spec:
+ requirements:
+   - key: karpenter.sh/capacity-type
+     operator: In
+     values: ["on-demand", "reserved"]
+```
+
+#### Launch ODCR instances with spot and on-demand fallback
+
+```yaml
+apiVersion: karpenter.sh/v1
+kind: NodePool
+metadata:
+ name: default
+spec:
+ # No additional requirements needed, launch all capacity types by default
+ requirements: []
+```
+
+## Scheduling Representation
+
+Since ODCRs are a AWS-specific concept, there needs to be a mechanism to pass down these ODCR options down for the scheduler to reason about. Importantly, we need the scheduler to know to prioritize these ODCR options when a user has specified them in their EC2NodeClass. Further, we need the scheduler to be aware that it can't launch an unlimited amount of these instances into an ODCR.
+
+_Note: Updates to the scheduling representation of our offerings, including changes to how the core scheduling behavior works are going to require extensive performance benchmarking to ensure that we do not significantly trade-off performance when we support users selecting on ODCRs._
+
+### Consider ODCR first during Scheduling
+
+Karpenter's current scheduling algorithm uses [First-Fit Decreasing bin-packing](https://en.wikipedia.org/wiki/First-fit-decreasing_bin_packing#:~:text=First%2Dfit%2Ddecreasing%20FFD,is%20at%20most%20the%20capacity). as a heuristic to optimize pod scheduling to nodes. For a new node that Karpenter chooses to launch, it will continue packing pods onto this new node until there are no more available instances type offerings. This happens regardless of the remaining capacity types in the offerings AND regardless of the price as offerings are removed.
+
+This presents a challenge for prioritizing ODCRs -- since this algorithm may remove `reserved` offerings to continue packing into `on-demand` and `spot` offerings, thus increasing the cost of the cluster and not fully utilizing the available capacity reservations.
+
+To solve for this problem, Karpenter will implement special handling for offerings under a specific price. If offerings are below a specific price threshold (e.g. `0.0000000001`), we will consider these offerings as "free" and uniquely prioritize them. This means that if we are about to remove an offering in our scheduling simulation that is below this price threshold such that there are no more offerings below this threshold, rather than scheduling this pod to the same node, we will create a new node, retaining the "free" offering, ensuring "free" offerings are prioritized by the scheduler.
```

Comment:
Interesting, so this is a hard-coded threshold? It feels like this should be some other attribute that signals that.

Reply:
Not intentionally hard-coded. I didn't mention that we would calculate this based off of the highest price instance type offering/lowest price instance type offering but that was the underlying idea here -- will update

Reply:
We could just have it in our offerings, treat it like normal in the scheduling code, and only reason about it in our Create() call?

Reply:
Discussed this offline: This prioritization is going to be driven by the `karpenter.sh/capacity-type: reserved` rather than through the pricing itself. This means that we will need upstream support for this concept and will have to move the feature gate upstream.

I'll create a separate design doc for this since this is going to be an upstream proposal.